### PR TITLE
Add support for cassandra #789

### PR DESCRIPTION
**Pull request by dawidmalina**

### CODE REVIEW

**Comment by dawidmalina** [comment]
cassandra

**Comment by emeroad** [comment]
Wow! Great~~ :+1:

**Comment by dawidmalina** [comment]
This is only for datastax driver and probably not cover all methods and statements. But at least we have something working:)

File: agent/src/main/resources/pinpoint-real-env-lowoverhead-sample.config [hunk 1]
```diff
@@ -167,6 +167,14 @@ profiler.jdbc.cubrid.rollback=false
 profiler.jdbc.dbcp=true
 profiler.jdbc.dbcp.connectionclose=false
 
+#
+# CASSANDRA
+#
+# Profile CASSANDRA.
+profiler.cassandra=true
```
Comment (lioolli) [inline agent/src/main/resources/pinpoint-real-env-lowoverhead-sample.config]: Plugin enable/disable is configured with `profiler.plugin.disable` property. You'd better remove this.

File: plugins/cassandra/src/main/java/com/navercorp/pinpoint/plugin/cassandra/CassandraConfig.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * @author dawidmalina
+ */
+public class CassandraConfig extends JdbcConfig {
+
+    private final boolean cassandra;
+
+    public CassandraConfig(ProfilerConfig config) {
+        super(config.readBoolean("profiler.jdbc.mysql.tracesqlbindvalue", config.isTraceSqlBindValue()), config.getMaxSqlBindValueSize());
```
Comment (lioolli) [inline plugins/cassandra/src/main/java/com/navercorp/pinpoint/plugin/cassandra/CassandraConfig.java]: maybe `profiler.cassandra.tracesqlbindvalue` not `profiler.jdbc.mysql.tracesqlbindvalue`, right?

**Comment by dawidmalina** [comment]
@lioolli Fix as you suggested. Additionally I've renamed sql to cql.

**Comment by Xylus** [comment]
@dawidmalina Could you rebase `dawidmalina:cassandra`? I'd like to work on the integration test off of it. Thanks!

**Comment by dawidmalina** [comment]
This is simple class I have used during plugin development (I hope will be helpful):
```
package com.navercorp.pinpoint.plugin.cassandra;

import com.datastax.driver.core.BoundStatement;
import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.PreparedStatement;
import com.datastax.driver.core.ResultSet;
import com.datastax.driver.core.Session;
import com.datastax.driver.core.Statement;
import com.datastax.driver.core.querybuilder.QueryBuilder;
import com.navercorp.pinpoint.bootstrap.plugin.test.PluginTestVerifier;
import com.navercorp.pinpoint.bootstrap.plugin.test.PluginTestVerifierHolder;
import com.navercorp.pinpoint.test.plugin.Dependency;
import com.navercorp.pinpoint.test.plugin.PinpointAgent;
import com.navercorp.pinpoint.test.plugin.PinpointPluginTestSuite;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.*;

@RunWith(PinpointPluginTestSuite.class)
@PinpointAgent
@Dependency({ "com.datastax.cassandra:cassandra-driver-core:2.1.7.1",
        "org.apache.cassandra:cassandra-all:2.1.9",
        "org.hamcrest:hamcrest-library:1.3"})
public class SimpleCQLTest {

    private static final String KEYSPACE = "mykeyspace";

    Cluster cluster;

    @Before
    public void setup() throws Exception {
        cluster = Cluster.builder().addContactPoints("192.168.99.100").withPort(9042).withoutMetrics().withoutJMXReporting()
                .build();
    }

    @Test
    public void testConnection() throws Exception {

        PluginTestVerifier verifier = PluginTestVerifierHolder.getInstance();
        Session session = cluster.connect(KEYSPACE);
        verifier.verifyTraceCount(1);

        ResultSet results = session.execute("SELECT * FROM mytable WHERE id='myKey01';");
        assertThat(results.iterator().next().getString("value"), is("myValue01"));
        verifier.printCache();
        verifier.verifyTraceCount(3);

        PreparedStatement statement = session.prepare("INSERT INTO myTable (id, value) VALUES (?,?);");
        BoundStatement boundStatement = new BoundStatement(statement);
        results = session.execute(boundStatement.bind("myKey03", "myValue03"));
        verifier.printCache();
        verifier.verifyTraceCount(5);

        Statement select = QueryBuilder.select().all().from(KEYSPACE, "mytable")
                .where(QueryBuilder.eq("id", "myKey03"));
        results = session.execute(select);
        assertEquals(1, results.all().size());
        verifier.printCache();
        verifier.verifyTraceCount(6);

        session.close();

    }

    public void cleanup() {
        cluster.close();
    }

}
```

**Comment by Xylus** [comment]
@dawidmalina Integration test looks good except for the class name, it should end with **IT**. (something like `SimpleCQLIT` or `CassandraClientIT`)
Also, checking for `SpanEvent` and their annotations would be even better.

Getting the actual Cassandra cluster seems like the problem. Have you considered starting up a `CassandraDaemon` for the tests? Kinda like how you started an embedded Maria DB instance for `MariaDBIT`.

**Comment by dawidmalina** [comment]
I had issue with embedded cassandra so I don't have anything I could share with you.

**Comment by Xylus** [comment]
Gotcha, I'll give it a try and let you know if anything comes out of it.

**Comment by bwzhang2011** [comment]
@dawidmalina, thanks a lot for such support for integration with cassandra as we know that zipkin backend storage is based on cassandra. so hope such could be tested and merged and released in next version. by the way,  cost for installing cassandra is much less thank hbase.

**Comment by bwzhang2011** [comment]
@Xylus, any update with such issue ?

**Comment by Xylus** [comment]
@bwzhang2011 working on integration test. Note this PR is for adding Datastax Cassandra client tracing - not adding storage support for Cassandra.

**Comment by bwzhang2011** [comment]
@Xylus, that's a bad news that I had thought that Cassandra could become the second choice for storage that I have to learn how to install hbase as we needed other work for hadoop installation. thanks for pointing out that.

**Comment by Xylus** [comment]
@dawidmalina Everything looks good, could you make one final squash + rebase so we can merge this?
Thanks

**Comment by dawidmalina** [comment]
@Xylus I have strange issue with squashing. Maybe we can merge it as is?

**Comment by dawidmalina** [comment]
I will create new PR

**Comment by dawidmalina** [comment]
This will be continued with: https://github.com/naver/pinpoint/pull/1630

**Pull request closed without merging**